arangod/Transaction/ManagerFeature.cpp

```diff
@@ -53,11 +52,9 @@ ManagerFeature::ManagerFeature(application_features::ApplicationServer& server)
     
     MANAGER->garbageCollect(/*abortAll*/false);
     
-    auto off = std::chrono::seconds(1);
-    
-    std::lock_guard<std::mutex> guard(_workItemMutex);
     if (!ApplicationServer::isStopping() && !canceled) {
-      _workItem = SchedulerFeature::SCHEDULER->queueDelay(RequestLane::INTERNAL_LOW, off, _gcfunc);
+      auto off = std::chrono::seconds(1);
+      std::atomic_store<Scheduler::WorkItem>(&_workItem, SchedulerFeature::SCHEDULER->queueDelay(RequestLane::INTERNAL_LOW, off, _gcfunc));
```

Comment:
I think it is different here.
The mutex (to which I reverted now) in `beginShutdown` and `stop` ensures that while it is held and the `_workItem` ptr is reset, the callback cannot reschedule itself at the same time.

So if `beginShutdown` or `stop` manage to acquire the mutex, they will destroy the workItem (which will cancel itself). This will call the workItem's callback function, but it will not reschedule itself, as we are already in the shutdown (ApplicationServer::isStopping() is guaranteed to return `true` here). So no rescheduling, and the callback is safe to execute then with either `canceled` set to true or not.
If `beginShutdown` or `stop` cannot acquire the mutex immediately, they will have to wait for the callback to either reschedule itself or to not reschedule itself. If it has rescheduled itself, `beginShutdown` or `stop` will cancel it afterwards. And if the callback has already found that the server is stopping, it will not have rescheduled it, and everything is fine already.

The ManagerFeature is still valid until the very end, so any of its internals can be accessed until after unprepare. And even its `MANAGER` member is valid until `unprepare`. So this should be ok too, as `beginShutdown` will only finish if there are no more transactions to garbage-collect.

Am I overlooking something that is unsafe in the mutex-based solution?

Reply:
You are right - I did not realize that in this case `beginShutdown` is not called from the destructor.
As long as it is guaranteed that the scheduler and all tasks are stopped at the time the ManagerFeature is destroyed this should be fine. And from your description and what I found in the code this seems to be the case.